### PR TITLE
[jasmine] Use type-safe spies for spyOnProperty

### PR DESCRIPTION
**Pull request by rslawik**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://jasmine.github.io/api/edge/global.html#spyOnProperty

### CODE REVIEW

**Review by amcasey — COMMENTED** [review]
LGTM, but I'd like to hear from an owner.

File: types/jasmine/index.d.ts [hunk 1]
```diff
@@ -191,7 +191,10 @@ declare function spyOn<T, K extends keyof T = keyof T>(
  * @param property The name of the property to replace with a `Spy`.
  * @param accessType The access type (get|set) of the property to `Spy` on.
  */
-declare function spyOnProperty<T>(object: T, property: keyof T, accessType?: "get" | "set"): jasmine.Spy;
+declare function spyOnProperty<T, K extends keyof T = keyof T>(
+    object: T, property: K, accessType?: "get"): jasmine.Spy<() => T[K]>;
```
Comment (SLaks) [inline types/jasmine/index.d.ts:195]: Both of these spy function types should probably also specify `this: T`

File: types/jasmine/jasmine-tests.ts [hunk 2]
```diff
@@ -1990,6 +1990,17 @@ describe("better typed spys", () => {
...
         });
     });
+    describe("spyOnProperty", () => {
+        it("works", () => {
+            const obj = {prop: "test", otherProp: 1};
+            const getSpy = spyOnProperty(obj, "prop");
+            getSpy.and.returnValue("spy");
+            getSpy.and.returnValue(123); // $ExpectError
+            const setSpy = spyOnProperty(obj, "prop", "set");
+            setSpy.calls.first().args; // $ExpectType [string] || [value: string]
```
Comment (chivesrs) [inline types/jasmine/jasmine-tests.ts:2000]: What exactly is the type on L2000? It doesn't make much sense to me

I get [string] is a 1 element tuple, but [value: string] I don't know?

File: types/jasmine/index.d.ts (hunk 1 above)
Comment (rslawik) [reply types/jasmine/index.d.ts]: Added `this: T`.

File: types/jasmine/jasmine-tests.ts (hunk 2 above)
Comment (rslawik) [reply types/jasmine/jasmine-tests.ts:2004]: Tuples in TypeScript may have labels: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-0.html#labeled-tuple-elements.

There is another example in L1984.

**Review by chivesrs — APPROVED** [review]

**Review by amcasey — APPROVED** [review]

**Review by SLaks — APPROVED** [review]

**Comment by rslawik** [comment]
Ready to merge

**Pull request merged**